proposals/p3762.md

```diff
@@ -0,0 +1,828 @@
+which defines an entity can omit `extern`, and omitting `extern` means it must
+provide a definition.
+
+An `extern` declaration forms an entity which has no definition or storage. For
+example:
+
+-   `extern fn` forms a function that can be called.
+-   `extern class` forms an incomplete type.
+-   `extern interface` forms an undefined interface.
+-   `extern var` or `extern let` will bind the name without allocating storage.
+    Initializers are disallowed.
+
+The `extern` keyword is invalid on declarations that have _only_ a declaration
+syntax and lack storage, such as `alias` or `namespace`. It is only valid on
+namespace-scoped names; it is invalid on type-scoped names
+(`class Foo { extern fn Member(); }`).
+
+In declaration modifiers, `extern` comes immediately after access control
+keywords, and before other keywords. For example,
+`private extern <other modifiers> class B;`. At present, when `extern` is on a
+declaration, only access modifiers are valid (see
+[Modifier keywords](#modifier-keywords)).
+
+### When declarations are allowed
+
+When considering whether a declaration is allowed, we apply the rules:
+
+1. A declaration should always add new information.
+    - No declarations after a definition.
+2. Only one library can declare an entity without `extern`.
+3. Support moving declarations between already-imported `api` files without
+   affecting compilation of client libraries.
+
+#### No forward declarations after declarations
+
+In a file, a forward declaration must never follow a forward declaration or
+definition for the same entity.
+
+For example:
+
+```carbon
+class A { ... }
+// Invalid: Disallowed after the definition.
+class A;
+
+class B;
+// Invalid: Disallowed due to repetition.
+class B:
+
+class C;
+// Valid: Allowed because the definition is added.
+class C { ... }
+```
+
+#### Files must either use an imported declaration or declare their own
+
+In a file, if a declaration or definition of an entity is imported, the file
+must choose between either using that version or declaring its own. It cannot do
+both.
+
+For example:
+
+```carbon
+package Foo library "a" api;
+
+class C { ... }
+```
+
+```carbon
+package Foo library "b" api;
+import library "a";
+
+extern class C;
+
+// Valid: Uses the incomplete type of the extern declaration.
+fn Foo(c: C*);
+```
+
+```carbon
+package Foo library "c" api;
+import library "a";
+
+// Valid: Uses the complete type of the imported definition.
+fn Foo(c: C);
+```
+
+```carbon
+package Foo library "d" api;
+import library "a";
+
+fn Foo(c: C);
+
+// Invalid: `Foo` used the imported `C`, so an `extern` declaration of `C` is now
+// invalid.
+extern class C;
+```
+
+This should have limited effect on `api` and `impl` combinations, but will still
+have some effect. For example:
+
+```carbon
+package Bar library "a" api;
+
+class C { ... }
+```
+
+```carbon
+package Bar library "a" impl;
+
+// Invalid: Imported the definition from the `api` file.
```

Comment:
Thanks for addressing this, I see you've moved the example next to where we say that case is invalid.

Just to confirm, is the case I mentioned in my previous comment valid? I don't feel strongly, but I think it makes sense to reject given that we reject in the case where the `api` file has a definition. The `api` file is "close enough" to the `impl` file that the declaration in the `impl` file seems like it should be rejected for being redundant.

Reply:
Published the response that I forgot to. Regarding the `api` case, let me give the more complex example of why I think it should be allowed:

```
package Baz library "a" api;
class C;
```

```
package Baz library "a" impl;
class C;
class D {
  fn E(c: C*) { ... }
}
class C {
  fn F(d: D) { ... }
}
```

If `class C;` is invalid, then `E` violates "Files must either use an imported declaration or declare their own". That can't just replace "Files" with "Libraries" because we already have an `api`/`impl` split in how that's handled, so there would be a finely tailored exception for this case. Rather, I think it's better to handle this way, allowing a marginally simpler rule set.

Reply:
I agree that we should allow your example, so my suggested rule generalization doesn't work.

But I think that doesn't answer the question about my example. To my reading, my example *is* disallowed by this proposal, specifically because it violates: "In an `impl` file, if the `impl` file forward declares an entity, it must also provide the definition." But yours is not disallowed, because the `impl` file provides the definition.

Does that match your expectation and intent?

Reply:
Per discussion, yes. Tried to make the example clearer.